qiskit/qpy/binary_io/circuits.py

```diff
@@ -578,7 +580,10 @@ def _dumps_instruction_parameter(param, index_map, use_symengine):
 
 # pylint: disable=too-many-boolean-expressions
 def _write_instruction(file_obj, instruction, custom_operations, index_map, use_symengine):
-    gate_class_name = instruction.operation.base_class.__name__
+    if isinstance(instruction.operation, Instruction):
+        gate_class_name = instruction.operation.base_class.__name__
+    else:
+        gate_class_name = instruction.operation.name
```

Comment:
I was a bit unsure about using `.name` here, since for circuit objects it can be pretty much anything. But since circuits will be caught with the above `if`-branch, I think using `.name` is fine, we just have to be careful for new operations. To be safer we _could_ use something like `type(instruction.operation).__name__` but that might be overkill.

Reply:
Alternatively, I could also change it to `instruction.operation.__class__.__name__ `, which is how is used to be before the singleton PR #11014. But I don't understand why we have to be careful or what's the problem that `.name` can be absolutely anything. Oh, you mean because you can create any gate, call it "clifford", and then the serializer will wrongly treat it as a Clifford gate? Or is the reason more involved than this?

Reply:
> Oh, you mean because you can create any gate, call it "clifford", and then the serializer will wrongly treat it as a Clifford gate?

Yeah exactly 🙂 That should be covered by the first if-clause but we're using `name` inconsistently here; for `Operations` we use the name attribute and for instructions the actual class name.